src/pip/_internal/index.py

```diff
@@ -59,8 +59,40 @@
 logger = logging.getLogger(__name__)
 
 
-def _get_content_type(url, session):
-    """Get the Content-Type of the given url, using a HEAD request"""
+def _match_vcs_scheme(url):
+    """Look for VCS schemes in the URL.
+
+    Returns the matched VCS scheme, or None if there's no match.
+    """
+    from pip._internal.vcs import VcsSupport
+    for scheme in VcsSupport.schemes:
+        if url.lower().startswith(scheme) and url[len(scheme)] in '+:':
+            return scheme
+    return None
+
+
+def _is_filename_like_archive(filename):
+    """Check whether the filename looks like an archive.
+    """
+    for bad_ext in ARCHIVE_EXTENSIONS:
+        if filename.endswith(bad_ext):
+            return True
+    return False
+
+
+class _NotHTML(Exception):
```

Comment:
Since `_NotHTML` can be raised in two different places, it seems like it might be good if the exception can distinguish those two cases (e.g. whether it's the result of a HEAD request vs. looking at the actual response). What about passing in a `desc` argument to `_ensure_html_header()` that would allow the two cases to be distinguished?

Reply:
What would the `desc` argument hold, the request type? (HEAD/GET)

Reply:
Information-wise, yes (at minimum). But keeping it more general (`desc` for description) allows it be any phrase with that information at no extra cost, e.g. "in response to HEAD."

Reply:
Ohhh, I failed to interpret `desc` as description somehow. It makes sense now 😅